contracts/profit/src/contract.rs

```diff
@@ -61,179 +63,91 @@ pub fn migrate(deps: DepsMut<'_>, _env: Env, _msg: MigrateMsg) -> ContractResult
 
 #[cfg_attr(feature = "contract-with-bindings", entry_point)]
 pub fn execute(
-    deps: DepsMut<'_>,
+    mut deps: DepsMut<'_>,
     env: Env,
     info: MessageInfo,
     msg: ExecuteMsg,
 ) -> ContractResult<CwResponse> {
     match msg {
         ExecuteMsg::TimeAlarm {} => {
+            SingleUserAccess::load(deps.storage, crate::access_control::TIMEALARMS_NAMESPACE)?
+                .check_access(&info.sender)?;
+
             let alarm_recepient = env.contract.address.clone();
-            try_transfer(deps.as_ref(), env, info).and_then(|resp| {
+
+            try_time_alarm(deps.branch(), env).and_then(|resp| {
                 response::response_with_messages::<_, _, ContractError>(&alarm_recepient, resp)
             })
         }
-    }
-}
+        ExecuteMsg::Config { cadence_hours } => {
+            SingleUserAccess::check_owner_access::<ContractError>(deps.storage, &info.sender)?;
 
-#[cfg_attr(feature = "contract-with-bindings", entry_point)]
-pub fn sudo(deps: DepsMut<'_>, _env: Env, msg: SudoMsg) -> ContractResult<CwResponse> {
-    match msg {
-        SudoMsg::Config { cadence_hours } => {
-            Profit::try_config(deps.storage, cadence_hours).map(|()| response::empty_response())
+            State::load(deps.storage)?
+                .try_update_config(cadence_hours)?
+                .store(deps.storage)?;
+
+            Ok(response::empty_response())
         }
     }
 }
 
 #[cfg_attr(feature = "contract-with-bindings", entry_point)]
-pub fn query(deps: Deps<'_>, _env: Env, msg: QueryMsg) -> ContractResult<Binary> {
-    match msg {
-        QueryMsg::Config {} => to_binary(&Profit::query_config(deps.storage)?),
-    }
-    .map_err(Into::into)
-}
+pub fn sudo(deps: DepsMut<'_>, env: Env, msg: NeutronSudoMsg) -> ContractResult<CwResponse> {
+    let state: State = State::load(deps.storage)?;
+
+    let DexResponse::<State> {
+        response,
+        next_state,
+    } = match msg {
+        NeutronSudoMsg::Response { data, .. } => state
+            .on_response(data, deps.as_ref(), env)
+            .continue_or_ok()?,
+        NeutronSudoMsg::Error { .. } => state.on_error(deps.as_ref(), env)?,
+        NeutronSudoMsg::Timeout { .. } => state.on_timeout(deps.as_ref(), env)?,
+        NeutronSudoMsg::OpenAck {
+            channel_id,
+            counterparty_channel_id,
+            counterparty_version,
+            ..
+        } => state.confirm_open(
+            deps.as_ref(),
+            env,
+            Ics20Channel {
+                local_endpoint: channel_id,
+                remote_endpoint: counterparty_channel_id,
+            },
+            counterparty_version,
+        )?,
```

Comment:
here we have to route the invocation to `DexSetupHandler` or `dex::Handler` depending on the input arguments, as follows:
- if `counterparty_version` is provided then call the latter, otherwise
- call the former mapping `port_id` as `ConnectionParams::connection_id`, `channel_id` as `Ics20Channel::local_endpoint`, and `counterparty_channel_id` as `Ics20Channel::remote_endpoint`

Reply:
Is that handler not called automatically by the `wasmd` modules?

Reply:
it is, with ICA account passed in `counterparty_version`. We'll call it via proposal when setting up DEX with out passing `counterparty_version`